### PR TITLE
Issue56510

### PR DESCRIPTION
**Pull request by JustABunchOfCode**
Add a custom XML-Parser, which was inspired by the Eclipse XML parser.  

https://issues.jenkins-ci.org/browse/JENKINS-56510

### CODE REVIEW

**Comment by uhafner** [comment]
Wasn't the idea to reuse the code in the warnings-plugin and the serialization of XStream?

**Comment by JustABunchOfCode** [comment]
The reason why i didn't reuse the code is: "Adding dependency on module 'warnings-ng' will introduce circular dependency between modules 'analysis-model' and 'warnings-ng'. "
Should I still reuse the code from warnings_ng?
Or should I only replace XPath (used by the EclipseXMLParser) with XStream?

**Comment by uhafner** [comment]
Well, actually the idea was to place the parser class into the module warnings-ng in order to simply things. But now that you have done all the work it is fine to leave it here. Typically converting a bean to an XML or JSON file is much simpler...

File: src/main/java/edu/hm/hafner/analysis/parser/xmlparser/XmlParser.java [hunk 1]
```diff
@@ -0,0 +1,206 @@
...
+
+    /**
+     * Read a value from XML and set a default value if its null.
+     * @param path = xpath.
+     * @param value = value which should be read.
+     * @param issue = current issue.
+     * @return value or default value.
+     * @throws XPathExpressionException error.
+     */
+    private String notNullableRead(final XPath path, final String value, final Element issue) throws XPathExpressionException {
```
Comment (uhafner) [inline src/main/java/edu/hm/hafner/analysis/parser/xmlparser/XmlParser.java]: I think it is save to return `null` here as the `IssueBuilder` instance correctly handles `null` values.

File: src/test/java/edu/hm/hafner/analysis/parser/xmlparser/XmlParserTest.java [hunk 2]
```diff
@@ -0,0 +1,145 @@
...
+                .hasSeverity(Severity.WARNING_LOW)
+                .hasMessage("Message 1")
+                .hasDescription("Description 1")
+                .hasPackageName("Package 1")
+                .hasModuleName("Module 1")
+                .hasOrigin("Origin 1")
+                .hasReference("Reference 1")
+                .hasFingerprint("Fingerprint 1")
+                .hasAdditionalProperties("Property 1, Property 2");
+
```
Comment (uhafner) [inline src/test/java/edu/hm/hafner/analysis/parser/xmlparser/XmlParserTest.java]: use auto format, just one blank line is enough

File: src/test/java/edu/hm/hafner/analysis/parser/xmlparser/XmlParserTest.java [hunk 3]
```diff
@@ -0,0 +1,145 @@
...
+    }
+
+    @Override
+    protected XmlParser createParser() {
+        return new XmlParser();
+    }
+
+    @Test
+    void shouldParseWithCustomPathAndMapper() {
```
Comment (uhafner) [inline src/test/java/edu/hm/hafner/analysis/parser/xmlparser/XmlParserTest.java]: Remove this test

File: src/test/java/edu/hm/hafner/analysis/parser/xmlparser/XmlParserTest.java [hunk 4]
```diff
@@ -0,0 +1,145 @@
...
+                    .hasOrigin("Origin 1")
+                    .hasReference("Reference 1")
+                    .hasFingerprint("Fingerprint 1")
+                    .hasAdditionalProperties("Property 1, Property 2");
+        });
+
+    }
+
+    @Test
+    void shouldThrowParserException() {
```
Comment (uhafner) [inline src/test/java/edu/hm/hafner/analysis/parser/xmlparser/XmlParserTest.java]: Can you add one test that shows that the parser still produces a list of issues if just one XML tag contains wrong values?

File: src/main/java/edu/hm/hafner/analysis/parser/xmlparser/XmlParser.java [hunk 5]
```diff
@@ -0,0 +1,206 @@
...
+                        .guessSeverity(notNullableRead(path, SEVERITY, issue))
+                        .setMessage(notNullableRead(path, MESSAGE, issue))
+                        .setDescription(notNullableRead(path, DESCRIPTION, issue))
+                        .setPackageName(notNullableRead(path, PACKAGE_NAME, issue))
+                        .setModuleName(notNullableRead(path, MODULE_NAME, issue))
+                        .setOrigin(notNullableRead(path, ORIGIN, issue))
+                        .setReference(notNullableRead(path, REFERENCE, issue))
+                        .setFingerprint(notNullableRead(path, FINGERPRINT, issue))
+                        .setAdditionalProperties(notNullableRead(path, ADDITIONAL_PROPERTIES, issue));
+                // TODO LineRange is missing.
```
[the remark this answers is not in the document]
Comment (JustABunchOfCode) [reply src/main/java/edu/hm/hafner/analysis/parser/xmlparser/XmlParser.java]: setLineRanges is used only five times (three times in tests) and has no "valueOfString"-Method which specifies String representation. Should I set a representation e.g. "start, end"? 

Also, there is no support for line ranges in IssueAssert. Is there a reason for that or should it be supplemented?

[the remark this answers is not in the document]
Comment (uhafner) [reply src/main/java/edu/hm/hafner/analysis/parser/xmlparser/XmlParser.java]: Line ranges are currently only used by SpotBugs/FindBugs and CPPLint. But they should be provided by the API so yes, we need these values. You can add any methods to `LineRange` and `LineRangeList` that are required to get this working. I updated [issue JENKINS-56510](https://issues.jenkins-ci.org/browse/JENKINS-56510) with an output of the `IssueStreamITest`, so you see how that field is serialized.

**Pull request merged**